aten/src/ATen/native/transformers/attention.cpp

```diff
@@ -387,6 +388,34 @@ std::tuple<Tensor, Tensor> native_multi_head_attention(
   const auto dim_per_head = D / num_head;
 #endif
 
+  const int64_t sdp_dim_per_head = D / num_head;
+  if ((query.is_same(key) && key.is_same(value)) && sdp_dim_per_head % 8 == 0 &&
+      query.is_cuda()) {
+    sdp::sdp_params kernel_params{
+        query, key, value, mask.has_value(), 0.0, need_weights, false};
+    auto backend = select_sdp_backend(kernel_params);
+    if (backend != sdp::SDPBackend::math && backend != sdp::SDPBackend::error) {
+      auto x = at::linear(query, qkv_weight, qkv_bias);
+      auto chunks = x.chunk(3, -1);
+      auto x_size_0 = x.size(0);
+
+      chunks[0] = (chunks[0].view({x_size_0, -1, num_head, sdp_dim_per_head}))
+                      .transpose(1, 2);
+      chunks[1] = (chunks[1].view({x_size_0, -1, num_head, sdp_dim_per_head}))
+                      .transpose(1, 2);
+      chunks[2] = (chunks[2].view({x_size_0, -1, num_head, sdp_dim_per_head}))
+                      .transpose(1, 2);
+
+      auto y = at::_scaled_dot_product_attention(
+          chunks[0], chunks[1], chunks[2], mask, 0.0, need_weights, false);
```

Comment:
Note that the last implementation from @danthe3rd allows to pass the tensor without having to reshape anything, avoiding a call to `contiguous` both before and after the operator. This can provide significant speedups btw